Cardinality Aggregation dynamic pruning null pointer exception

We got a report of null pointer exception when performing dynamic pruning in cardinality aggregation.

The exception looks like below

```
Caused by: java.lang.NullPointerException: Cannot read field \"doc\" because \"top\" is null
at org.opensearch.search.aggregations.metrics.CardinalityAggregator$PruningCollector.prune(CardinalityAggregator.java:379)
at org.opensearch.search.aggregations.metrics.CardinalityAggregator$PruningCollector.collect(CardinalityAggregator.java:371)
at org.opensearch.search.aggregations.LeafBucketCollector.collect(LeafBucketCollector.java:123)
at org.apache.lucene.search.BooleanScorer$DocIdStreamView.forEach(BooleanScorer.java:178)
at org.apache.lucene.search.LeafCollector.collect(LeafCollector.java:106)
at org.apache.lucene.search.BooleanScorer.scoreWindowIntoBitSetAndReplay(BooleanScorer.java:257)
at org.apache.lucene.search.BooleanScorer.scoreWindowMultipleScorers(BooleanScorer.java:309)
at org.apache.lucene.search.BooleanScorer.scoreWindow(BooleanScorer.java:362)
at org.apache.lucene.search.BooleanScorer.score(BooleanScorer.java:374)
at org.apache.lucene.search.BulkScorer.score(BulkScorer.java:38)
at org.opensearch.search.aggregations.metrics.CardinalityAggregator.tryScoreWithPruningCollector(CardinalityAggregator.java:229)\n\t... 41 more\n
```

Notice the code path starts from BooleanScorer. This is missed during development where we focused mostly just on DefaultBulkScorer. Although DefaultBulkScorer guarantee no NPE at that place, but the other implementations of BulkScorer may not have that guarantees. 

The fix is to simply add a null check.

--- a/server/src/main/java/org/opensearch/search/aggregations/metrics/CardinalityAggregator.java
+++ b/server/src/main/java/org/opensearch/search/aggregations/metrics/CardinalityAggregator.java
@@ -382,7 +382,7 @@ public class CardinalityAggregator extends NumericMetricsAggregator.SingleValue 
          */
         private void prune(int doc) {
             if (queue.size() == 0) {
-                return ;
+                return;
             }
             DisiWrapper top = queue.top();
             if (top == null) {

--- a/server/src/main/java/org/opensearch/search/aggregations/metrics/CardinalityAggregator.java
+++ b/server/src/main/java/org/opensearch/search/aggregations/metrics/CardinalityAggregator.java
@@ -382,7 +382,7 @@ public class CardinalityAggregator extends NumericMetricsAggregator.SingleValue 
          */
         private void prune(int doc) {
             if (queue.size() == 0) {
-                return;
+                return ;
             }
             DisiWrapper top = queue.top();
             if (top == null) {

--- a/server/src/main/java/org/opensearch/search/aggregations/metrics/CardinalityAggregator.java
+++ b/server/src/main/java/org/opensearch/search/aggregations/metrics/CardinalityAggregator.java
@@ -382,11 +382,11 @@ public class CardinalityAggregator extends NumericMetricsAggregator.SingleValue 
          */
         private void prune(int doc) {
             if (queue.size() == 0) {
-                return ;
+                return;
             }
             DisiWrapper top = queue.top();
             if (top == null) {
-                return ;
+                return;
             }
             int curTopDoc = top.doc;
             if (curTopDoc == doc) {

--- a/server/src/main/java/org/opensearch/search/aggregations/metrics/CardinalityAggregator.java
+++ b/server/src/main/java/org/opensearch/search/aggregations/metrics/CardinalityAggregator.java
@@ -381,7 +381,13 @@ public class CardinalityAggregator extends NumericMetricsAggregator.SingleValue 
          * Note: the queue may be empty or the queue top may be null after pruning
          */
         private void prune(int doc) {
+            if (queue.size() == 0) {
+                return ;
+            }
             DisiWrapper top = queue.top();
+            if (top == null) {
+                return ;
+            }
             int curTopDoc = top.doc;
             if (curTopDoc == doc) {
                 do {

